pkg/component/runtime/log_writer.go

```diff
@@ -0,0 +1,171 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package runtime
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/elastic/elastic-agent/pkg/component"
+)
+
+type zapcoreWriter interface {
+	Write(zapcore.Entry, []zapcore.Field) error
+}
+
+// logWriter is an `io.Writer` that takes lines and passes them through the logger.
+//
+// `Write` handles parsing lines as either ndjson or plain text.
+type logWriter struct {
+	loggerCore zapcoreWriter
+	logCfg     component.CommandLogSpec
+	remainder  []byte
+}
+
+func newLogWriter(core zapcoreWriter, logCfg component.CommandLogSpec) *logWriter {
+	return &logWriter{
+		loggerCore: core,
+		logCfg:     logCfg,
+	}
+}
+
+func (r *logWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		// nothing to do
+		return 0, nil
+	}
+	offset := 0
+	for {
+		idx := bytes.IndexByte(p[offset:], '\n')
```

Comment:
i'm ok with this, i imagine you considered bytes.SplitN. 
memory consumption would be higher but implementation easier. indexByte also scans whole string for a match so there won't be cpu overhead.